Add Go compiler directive 'noescape' to assembly implementations

smira opened the pull request:
Without this directive Go considers that `data []byte` escapes from the
function, and forces allocation on the heap even for values which can
perfectly stay on the stack.

This removes extra allocation when value passed in is not on the heap
already.

See also: https://github.com/golang/go/issues/4099

twmb commented:
Tip again down to negligible improvements; I'll evaluate this when 1.7 is released.

twmb commented:
Tip, today, with larger 32 and 128 cmps:

```
benchmark                old ns/op     new ns/op     delta
Benchmark32_1-2          7.89          6.48          -17.87%
Benchmark32_2-2          9.11          6.48          -28.87%
Benchmark32_4-2          8.47          6.53          -22.90%
Benchmark32_8-2          10.7          8.61          -19.53%
Benchmark32_16-2         13.6          10.9          -19.85%
Benchmark32_32-2         20.6          17.5          -15.05%
Benchmark32_64-2         35.2          27.9          -20.74%
Benchmark32_128-2        67.1          57.6          -14.16%
Benchmark32_256-2        131           116           -11.45%
Benchmark32_512-2        249           217           -12.85%
Benchmark32_1024-2       510           426           -16.47%
Benchmark32_2048-2       1034          819           -20.79%
Benchmark32_4096-2       1976          1651          -16.45%
Benchmark32_8192-2       4155          3334          -19.76%
Benchmark32_16384-2      8143          6653          -18.30%
Benchmark32_32768-2      16155         13129         -18.73%
Benchmark32_65536-2      32509         28332         -12.85%
Benchmark128_1-2         23.8          11.4          -52.10%
Benchmark128_2-2         25.9          12.1          -53.28%
Benchmark128_4-2         26.1          12.4          -52.49%
Benchmark128_8-2         26.4          14.5          -45.08%
Benchmark128_16-2        24.3          13.3          -45.27%
Benchmark128_32-2        25.8          17.7          -31.40%
Benchmark128_64-2        33.1          22.9          -30.82%
Benchmark128_128-2       45.4          35.0          -22.91%
Benchmark128_256-2       72.1          59.2          -17.89%
Benchmark128_512-2       129           102           -20.93%
Benchmark128_1024-2      222           191           -13.96%
Benchmark128_2048-2      408           352           -13.73%
Benchmark128_4096-2      819           680           -16.97%
Benchmark128_8192-2      1656          1417          -14.43%
Benchmark128_16384-2     3359          2795          -16.79%
Benchmark128_32768-2     6594          6080          -7.79%
Benchmark128_65536-2     13357         11560         -13.45%
```

On a 2010 Mac (tip):
```
Benchmark32_1-4          7.68          6.52          -15.10%
Benchmark32_2-4          8.76          6.90          -21.23%
Benchmark32_4-4          8.31          6.84          -17.69%
Benchmark32_8-4          10.0          8.10          -19.00%
Benchmark32_16-4         13.2          11.0          -16.67%
Benchmark32_32-4         20.7          16.6          -19.81%
Benchmark32_64-4         36.7          31.0          -15.53%
Benchmark32_128-4        76.5          55.3          -27.71%
Benchmark32_256-4        153           122           -20.26%
Benchmark32_512-4        310           250           -19.35%
Benchmark32_1024-4       630           500           -20.63%
Benchmark32_2048-4       1255          1021          -18.65%
Benchmark32_4096-4       2519          2033          -19.29%
Benchmark32_8192-4       5039          4192          -16.81%
Benchmark32_16384-4      9978          8182          -18.00%
Benchmark32_32678-4      20160         16312         -19.09%
Benchmark32_65536-4      40328         32478         -19.47%
Benchmark128_1-4         22.1          11.2          -49.32%
Benchmark128_2-4         23.5          11.6          -50.64%
Benchmark128_4-4         24.2          12.3          -49.17%
Benchmark128_8-4         26.5          13.0          -50.94%
Benchmark128_16-4        22.6          13.0          -42.48%
Benchmark128_32-4        25.5          15.0          -41.18%
Benchmark128_64-4        32.3          20.8          -35.60%
Benchmark128_128-4       45.4          35.0          -22.91%
Benchmark128_256-4       70.7          59.0          -16.55%
Benchmark128_512-4       123           110           -10.57%
Benchmark128_1024-4      241           211           -12.45%
Benchmark128_2048-4      443           415           -6.32%
Benchmark128_4096-4      861           817           -5.11%
Benchmark128_8192-4      1727          1656          -4.11%
Benchmark128_16384-4     3406          3271          -3.96%
Benchmark128_32678-4     6802          6567          -3.45%
Benchmark128_65536-4     13571         13034         -3.96%
```

smira committed:
Add Go compiler directive 'noescape' to assembly implementations

Without this directive Go considers that `data []byte` escapes from the
function, and forces allocation on the heap even for values which can
perfectly stay on the stack.

This removes extra allocation when value passed in is not on the heap
already.

See also: https://github.com/golang/go/issues/4099
